File: day-12/main.rb
# Advent of Code 2015 - Day 12

require "json"

def read_input
  File.read(File.expand_path("../input.json", __FILE__))
end

def recursive_sum(item)
  return 0 if item.is_a?(Hash) && item.value?("red")

  case item
  when Array
    item.map { |value| recursive_sum(value) }.reduce(0, :+)
  when Fixnum
    item
  when Hash
    item.map { |_key, value| recursive_sum(value) }.reduce(0, :+)
  else
    0
  end
end

def main
  json = read_input

  number_sum = json.scan(/[+-]?\d+/).map(&:to_i).reduce(0, :+)
  nonred_sum = recursive_sum(JSON.parse(json))

  puts "Part One: #{number_sum}"
  puts "Part Two: #{nonred_sum}"
end

main if "main" == to_s

File: day-12/main.py
"""Advent of Code 2015 - Day 12."""

import json
import numbers
from pathlib import Path


def read_input():
    """Read input file."""
    return Path(__file__).with_name('input.json').read_text()


def traverse_json(data, filter=None):
    """Recursively traverse JSON structure yielding all its elements."""
    if filter and filter(data):
        return
    yield data

    # Handle nested structures like dicts/lists and yield their values, too.
    if isinstance(data, dict):
        for item in data.values():
            yield from traverse_json(item, filter=filter)
    elif isinstance(data, list):
        for item in data:
            yield from traverse_json(item, filter=filter)


def is_number(item):
    """Check if the item is a number."""
    return isinstance(item, numbers.Number)


def is_red_dict(item):
    """Check if the item is a dict that contains a red value."""
    return isinstance(item, dict) and 'red' in item.values()


def main():
    """Main entry point of puzzle solution."""
    data = json.loads(read_input())

    part_one = sum(filter(is_number, traverse_json(data)))
    part_two = sum(filter(is_number, traverse_json(data, filter=is_red_dict)))

    print('Part One: {}'.format(part_one))
    print('Part Two: {}'.format(part_two))


if __name__ == '__main__':
    main()
